### PR TITLE
Added params_require_grad arg to make_functional*

### PR DESCRIPTION
**Pull request by vfdev-5**
Related to https://github.com/pytorch/functorch/issues/639#issuecomment-1097154660

Description:
- Added `params_require_grad` argument to `make_functional` and `make_functional_with_buffers`
- Added tests

@samdow let me know if something else I'm missing for this PR. Thanks

### CODE REVIEW

File: functorch/_src/make_functional.py [hunk 1]
```diff
@@ -310,7 +316,7 @@ def forward(self, params, *args, **kwargs):
             _swap_state(self.stateless_model, self.names_map, old_state)
 
 
-def make_functional(model: nn.Module):
+def make_functional(model: nn.Module, params_require_grad: bool = True):
```
Comment (zou3519) [inline functorch/_src/make_functional.py:319]: I have one main question about how this API works:
1. params_require_grad: should we make all parameters require grad, or should we mirror the state on the nn.Module (can a parameter on nn.Module have requires_grad=False) ?

Comment (vfdev-5) [reply functorch/_src/make_functional.py:319]: Yes, I agree, maybe this API can be confusing at this point. Probably, naming should reflect the fact that we could disable enabled grads but not enable disabled... Let me think about the naming.

Comment (samdow) [reply functorch/_src/make_functional.py:319]: Agree that we can fix this with naming, especially since @vfdev-5 (correctly) only sets the params to `requires_grad(False)` and never the other way around. Maybe (or at least as a jumping off point),`disable_params_grad` (flipping the boolean's default)?

File: functorch/_src/make_functional.py [hunk 2]
```diff
@@ -353,15 +359,21 @@ def compute_loss(params, x, t):
 
     If the model has any buffers, please use :func:`make_functional_with_buffers` instead.
 
+    Args:
+        model (torch.nn.Module): Input model.
+        params_require_grad (bool): If True, output parameters will have require gradient.
+            It could be helpful to set this flag to False if computions do not need higher order gradients.
```
Comment (samdow) [inline functorch/_src/make_functional.py:365]: nit: maybe "computations do not need higher order gradients" -> "gradient computations only use the functorch API"

This point definitely got jumbled in the discussion on the issue, but basically we can still get higher order gradients with this, we just would have to use functorch's `grad(grad(...))`. The issue will be if we did something like `grad(...).backward()`

Comment (vfdev-5) [reply functorch/_src/make_functional.py]: @samdow could you please check again the docstring and maybe help to improve it.

File: functorch/_src/make_functional.py (hunk 1 above)
Comment (vfdev-5) [reply functorch/_src/make_functional.py]: Let's make it as `disable_params_grad`

File: functorch/_src/make_functional.py [hunk 3]
```diff
@@ -353,15 +359,22 @@ def compute_loss(params, x, t):
 
     If the model has any buffers, please use :func:`make_functional_with_buffers` instead.
 
+    Args:
+        model (torch.nn.Module): Input model.
+        disable_params_grad (bool): Flag to disable gradients for output parameters.
+            It could be helpful to set this flag to True if gradient computations only use the functorch API.
+            Disabling gradients also helps to avoid accumulating them in the backprop graph and to prevent
+            keeping them in memory. Default: False.
```
Comment (zou3519) [inline functorch/_src/make_functional.py:367]: This is a bit ambiguous. I'd phrase it something like the following:
- The returned params are unrelated to the set of params from the original model. If True (default), the params will have requires_grad=True on them (aka they will be trackable with regular PyTorch autograd), matching the requires_grad-ness of the params from the original model. Otherwise, the returned params will have requires_grad=False.
- If you plan on using regular PyTorch autograd (e.g., if you want to call .backward() or torch.autograd.grad(), then set disable_params_grad=True (default)). Otherwise, if you're only planning on using functorch's gradient transforms, then please set disable_params_grad=False to avoid unnecessarily tracking history with PyTorch autograd.

File: functorch/_src/make_functional.py [hunk 4]
```diff
@@ -353,15 +359,22 @@ def compute_loss(params, x, t):
 
     If the model has any buffers, please use :func:`make_functional_with_buffers` instead.
 
+    Args:
+        model (torch.nn.Module): Input model.
+        disable_params_grad (bool): Flag to disable gradients for output parameters.
```
Comment (zou3519) [inline functorch/_src/make_functional.py:364]: Can we bikeshed this name a little more? The reason why it is confusing for me is that we're not really disabling gradients, what we are doing is turning on or off regular PyTorch Autograd tracking of the input Tensors.

Here are some potential names:
- set_requires_grad (requires_grad is the property that PyTorch autograd uses to track gradients)
- requires_grad
- enable_pytorch_autograd

Thoughts? @vfdev-5 @samdow

Comment (samdow) [reply functorch/_src/make_functional.py:364]: Per the other discussion, I think `set_requires_grad ` and `requires_grad ` might be confusing since parameters don't _have_ to `requires_grad` and we aren't changing them if it's True

Maybe `enable_pytorch_autograd`? @vfdev-5 what do you think?

Comment (zou3519) [reply functorch/_src/make_functional.py:364]: > disable_autograd_tracking

I like this better. Eventually we'll upstream functorch into pytorch/pytorch and "enable_pytorch_autograd" will get even more confusing: "What do you mean by PyTorch autograd?"

Comment (samdow) [reply functorch/_src/make_functional.py:364]: Same! Thanks for the suggestion @vfdev-5!

**Comment by vfdev-5** [comment]
@zou3519 @samdow I updated the name and the docstring (and skipped the CI) to wait for your approval. If everything is fine, then I can restart the CI, we'll be able to merge.

File: functorch/_src/make_functional.py [hunk 5]
```diff
@@ -310,7 +316,7 @@ def forward(self, params, *args, **kwargs):
             _swap_state(self.stateless_model, self.names_map, old_state)
 
 
-def make_functional(model: nn.Module):
+def make_functional(model: nn.Module, disable_autograd_tracking: bool = False):
     """make_functional(model) -> func, params
```
Comment (samdow) [inline functorch/_src/make_functional.py:320]: nit: (sorry I missed this earlier) We should update this to match the new signature

File: functorch/_src/make_functional.py [hunk 6]
```diff
@@ -353,15 +359,28 @@ def compute_loss(params, x, t):
...
     if len(buffers) > 0:
         raise RuntimeError('make_functional(model): `model` has buffers. Please use '
                            'make_functional_with_buffers(model) instead.')
-    return FunctionalModule._create_from(model)
+    return FunctionalModule._create_from(model, disable_autograd_tracking=disable_autograd_tracking)
 
 
-def make_functional_with_buffers(model: nn.Module):
+def make_functional_with_buffers(model: nn.Module, disable_autograd_tracking: bool = False):
     """make_functional_with_buffers(model) -> func, params, buffers
```
Comment (samdow) [inline functorch/_src/make_functional.py:384]: nit: ditto here

**Review by samdow — APPROVED** [review]
LGTM! Just had the small nit that I missed earlier. Thanks for the patience with the doc + naming comments 😄

File: functorch/_src/make_functional.py (hunk 5 above)
Comment (vfdev-5) [reply functorch/_src/make_functional.py:320]: @samdow sorry for delay. Actually, I didn't change this part of docstring as `disable_autograd_tracking` arg is not essential and IMO `make_functional(model) -> func, params` is showing essential usage.
What do you think ? I can change it though if you confirm. Thanks

Comment (vfdev-5) [reply functorch/_src/make_functional.py:320]: Oh, I see, I haven't remarked that previously. Thanks! Let me update the docstring here as well

**Comment by samdow** [comment]
Merging since all of the failures look to be from upstream. Thanks for all your work on this @vfdev-5!
